fix(metrics/prometheus): gracefully handle multiple event sources, avoid erroneous duplicate metrics

incertum opened the pull request:
<!--  Thanks for sending a pull request! Here are some tips for you:
1. If this is your first time, please read our contributor guidelines in the https://github.com/falcosecurity/.github/blob/main/CONTRIBUTING.md file.
2. Please label this pull request according to what type of issue you are addressing.
3. Please add a release note!
4. If the PR is unfinished while opening it specify a wip in the title before the actual title, for example, "wip: my awesome feature"
-->

**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

/kind bug

/kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

> /kind feature

> /kind release

<!--
Please remove the leading whitespace before the `/kind <>` you uncommented.
-->

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area build

/area engine

> /area tests

> /area proposals

> /area CI

<!--
Please remove the leading whitespace before the `/area <>` you uncommented.
-->

**What this PR does / why we need it**:

Gracefully handle multiple event sources, avoid erroneous duplicate metrics.
Issue found and reported by @norman-ziebal, thank you!

**Which issue(s) this PR fixes**:

https://github.com/falcosecurity/falco/issues/3539

<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
If PR is `kind/failing-tests` or `kind/flaky-test`, please post the related issues/tests in a comment and do not use `Fixes`.
-->

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

<!--
If NO, just write "NONE" in the release-note block below.

If YES, a release note is required, enter your release note in the block below. 
The convention is the same as for commit messages: https://github.com/falcosecurity/.github/blob/main/CONTRIBUTING.md#commit-convention
If the PR introduces non-backward compatible changes, please add a line starting with "BREAKING CHANGE:" and describe what changed.
For example, `BREAKING CHANGE: the API interface of the rule engine has changed`.
Your note will be included in the changelog.
-->

```release-note
fix(metrics/prometheus): gracefully handle multiple event sources, avoid erroneous duplicate metrics
```

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/falco/pull/3563#" title="Author self-approved">incertum</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Ffalco).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/falco/blob/master/OWNERS)~~ [incertum]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

incertum commented:
/milestone 0.41.0

incertum commented on userspace/falco/falco_metrics.cpp:
```diff
@@ -72,9 +72,22 @@ std::string falco_metrics::to_text(const falco::app::state& state) {
+		// sources
 		//
+
+		/* Examples ...
```
@sgaist this seems to be the only true reason for why we would need to loop over each inspector vs only needing the first one.  The examples listed here were verified by me when testing with the dummy_c event source plugin.

incertum commented on userspace/falco/falco_metrics.cpp:
```diff
@@ -86,306 +99,346 @@ std::string falco_metrics::to_text(const falco::app::state& state) {
+
+			// Each inspector includes all event sources
+			/* Examples ...
+			    # HELP falcosecurity_falco_evt_source_info https://falco.org/docs/metrics/
```
@sgaist and @FedeDP: We seem to have them in each inspector, hence we may only need to query the first inspector. While testing with the dummy_c evert sourc plugin this assumptions appears to hold true.

incertum commented on userspace/falco/falco_metrics.cpp:
```diff
@@ -86,306 +99,346 @@ std::string falco_metrics::to_text(const falco::app::state& state) {
-			        "evt",
-			        {{"hostname", machine_info->hostname}});
-		}
+		if(i == 0) {
```
This causes a larger diff and will make the PR more cumbersome to review given the indentation changes, sorry about that ...

incertum commented on userspace/falco/falco_metrics.cpp:
```diff
@@ -86,306 +99,346 @@ std::string falco_metrics::to_text(const falco::app::state& state) {
+			// Performed repeatedly for each inspectors' libs metrics collector
+
+			for(auto& metric : metrics_snapshot) {
+				if(metric.flags & METRICS_V2_PLUGINS) {
```
This one I didn't explicitly test or could test at this time, but I assume that for each inspector we should loop over the libs metrics collectors' plugin metrics ...

However, I can share that the existing metrics and the metrics from the new container plugin are showing up, so it should at the very least not break anything.

incertum commented on userspace/falco/falco_metrics.cpp:
```diff
@@ -86,6 +99,31 @@ std::string falco_metrics::to_text(const falco::app::state& state) {
+		//
+
+		// Each inspector includes all event sources
+		/* Examples ...
```
@sgaist and @FedeDP: We seem to have them in each inspector, hence we may only need to query the first inspector. While testing with the dummy_c evert sourc plugin this assumptions appears to hold true.

FedeDP replied:
This is really weird since the plugins get registered only to their source inspector (and in the offline inspector):
```
plugin = src_info->inspector->register_plugin(config->m_library_path);
```
See https://github.com/falcosecurity/falco/blob/145036e92369d9ae5a6ced848bdeb1d0b0fbafd5/userspace/falco/app/actions/init_inspectors.cpp#L137

incertum replied:
Yes gremlins again, see the output now, I moved it back to the loop

```
# HELP falcosecurity_falco_evt_source_info https://falco.org/docs/metrics/
# TYPE falcosecurity_falco_evt_source_info gauge
falcosecurity_falco_evt_source_info{evt_source="syscall"} 1

...

# HELP falcosecurity_falco_evt_source_info https://falco.org/docs/metrics/
# TYPE falcosecurity_falco_evt_source_info gauge
falcosecurity_falco_evt_source_info{evt_source="syscall"} 1
# HELP falcosecurity_falco_evt_source_info https://falco.org/docs/metrics/
# TYPE falcosecurity_falco_evt_source_info gauge
falcosecurity_falco_evt_source_info{evt_source="dummy_c"} 1
```

incertum commented on userspace/falco/falco_metrics.cpp:
```diff
@@ -285,18 +316,42 @@ std::string falco_metrics::to_text(const falco::app::state& state) {
 		auto metrics_snapshot = metrics_collector.get_metrics();
 
+		if(i != 0) {
+			// Performed repeatedly for each inspectors' libs metrics collector
```
This one I didn't explicitly test or could test at this time, but I assume that for each inspector we should loop over the libs metrics collectors' plugin metrics ...

However, I can share that the existing metrics and the metrics from the new container plugin are showing up, so it should at the very least not break anything.

incertum commented on userspace/falco/falco_metrics.cpp:
```diff
@@ -72,9 +72,22 @@ std::string falco_metrics::to_text(const falco::app::state& state) {
+		// sources
 		//
+
+		/* Examples ...
```
@sgaist this seems to be the only true reason for why we would need to loop over each inspector vs only needing the first one. The examples listed here were verified by me when testing with the dummy_c event source plugin.

FedeDP replied:
List of inspectors is created above (L64) from the state `enabled_sources`, thus we are responsible to eventually put the `syscall` inspector first in the vector :)

FedeDP replied:
Also, question: what if we have multiple source plugin enabled?
Won't
> falcosecurity_scap_engine_name_info{engine_name="source_plugin"} 1

be pushed multiple times (causing the issue that we trying to fix?)

What if we instead pushed the number of inspectors that use a certain engines, like:
```
falcosecurity_scap_engine_name_info{engine_name="source_plugin"} 4
falcosecurity_scap_engine_name_info{engine_name="bpf"} 1
```
?

incertum replied:
re the comment around ordering, and having the syscalls inspector at index 0 - I adjusted the approach, seems to work now.

re multiple plugin event sources -- from my perspective maybe more of a tomorrow problem and not highest priority since these less essential informational metrics compared to the more important metrics.

re not having the syscalls inspector at all, then any of the other inspectors is at index 0 and will do it's best to fetch as many metrics as it can -- all we can guarantee at this time. We did some fixes in libs a while ago, such as adding a linux lite platform so that at the very least `agent_info` and `machine_info` are available for plugin even sources only (without needing the syscalls event source).

incertum commented:
> I fear we will need multiple iterations on this PR :/ First of all, i'd try to understand which are the metrics that we need to push multiple times, ie: once for each enabled source (inspector). For example, i think that both `agent_info` and `machine_info` can only be pushed once; same for Falco version and probably loaded rules and configs.
> 
> Once we understand what needs to be pushed specifically for each source, we can go ahead!
> 
> I'll try a first pass:
> 
>     * as suggested below, `engine_name` metric should be pushed with the counter of enabled engines, for each engine type (ie: looping over inspectors from enabled sources)
> 
>     * agent_info, machine_info, Falco version, config and rules shasums, output queues drops, additional_wrapper_metrics are inspector-agnostic (even if agent_info and machine_info need an inspector; will need to double check if any inspector is good for them)
> 
>     * `evt_source` metric should be pushed for each inspector (? i guess, not sure here), perhaps adding the engine name as additional label
> 
>     * rule_counters should be pushed specifically for each source
> 
>     * jemalloc stats are agnostic
> 
>     * libs metrics collector are specific for each inspector

@FedeDP first I added a commit to ensure the syscalls inspector is always the first one. Now everything seems correct. There is also an issue in this regard wrt the metrics rule, but we won't touch that for now - that would be crazy ...

Subsequently, I added a co-authored comment, and I attempted to utilize your reasoning provided above plus my 2 cents, WDYT?

Re timing, yes we can push it to the next release or consider it for a patch release or if you now believe we could still make it I am here to support getting it over the finsih line!

incertum commented:
@FedeDP thanks for merging the first metrics related PR :heart:. 

I have quickly rebased this PR so that you can start working on it. Fede and I discussed that we are now doing a handoff since I believe it is now fixed for the Prometheus metrics, and Fede will confirm if anything else may be needed and if my claims hold true ;) @sgaist if you have time to help us, we would truly appreciate it! Plus if you both happen to have ideas re how to make the code in `falco_metrics` more organized given these changes, it would also be much appreciated!

(@FedeDP when you run this with the `dummy_c` plugin event source and check out the output rules metrics you'll directly see that there is yet another issue that I have not yet addressed (only left a comment). Perhaps you can take a look and add a second opinion around whether we should defer that fix to the next release?)

FedeDP commented:
Pushed some changes:
* refactored the `to_text` method to be cleaner
* we need to only push metrics for enabled sources
* refactored code a bit
* we will only push agent_info and machine_info for first source inspector that expsoes them

Example prometheus metrics:
```
curl http://127.0.0.1:8765/metrics
# HELP falcosecurity_falco_version_info https://falco.org/docs/metrics/
# TYPE falcosecurity_falco_version_info gauge
falcosecurity_falco_version_info{version="0.41.0-100+334ca42"} 1
# HELP falcosecurity_falco_sha256_rules_files_info https://falco.org/docs/metrics/
# TYPE falcosecurity_falco_sha256_rules_files_info gauge
falcosecurity_falco_sha256_rules_files_info{file_name="falco_rules.yaml",sha256="6f0078862a26528cb50a860f9ebebbfbe3162e5009187089c73cb0cdf91d0b06"} 1
# HELP falcosecurity_falco_sha256_config_files_info https://falco.org/docs/metrics/
# TYPE falcosecurity_falco_sha256_config_files_info gauge
falcosecurity_falco_sha256_config_files_info{file_name="falco.yaml",sha256="5b9f83bc0aa52a9ed36bbdddcef0cc6236a8c8e111749a79ceefe0e8c60bffc6"} 1
# HELP falcosecurity_falco_rules_matches_total https://falco.org/docs/metrics/
# TYPE falcosecurity_falco_rules_matches_total counter
falcosecurity_falco_rules_matches_total{priority="4",rule_name="Read sensitive file untrusted",source="syscall",tag_T1555="true",tag_container="true",tag_filesystem="true",tag_host="true",tag_maturity_stable="true",tag_mitre_credential_access="true"} 32
# HELP falcosecurity_falco_rules_matches_total https://falco.org/docs/metrics/
# TYPE falcosecurity_falco_rules_matches_total counter
falcosecurity_falco_rules_matches_total{priority="5",rule_name="Terminal shell in container",source="syscall",tag_T1059="true",tag_container="true",tag_maturity_stable="true",tag_mitre_execution="true",tag_shell="true"} 1
# HELP falcosecurity_scap_engine_name_info https://falco.org/docs/metrics/
# TYPE falcosecurity_scap_engine_name_info gauge
falcosecurity_scap_engine_name_info{engine_name="bpf",evt_source="syscall"} 1
# HELP falcosecurity_scap_n_evts_total https://falco.org/docs/metrics/
# TYPE falcosecurity_scap_n_evts_total counter
falcosecurity_scap_n_evts_total 320384
# HELP falcosecurity_scap_n_drops_buffer_total https://falco.org/docs/metrics/
# TYPE falcosecurity_scap_n_drops_buffer_total counter
falcosecurity_scap_n_drops_buffer_total{dir="enter",drop="clone_fork"} 0
# HELP falcosecurity_scap_n_drops_buffer_total https://falco.org/docs/metrics/
# TYPE falcosecurity_scap_n_drops_buffer_total counter
falcosecurity_scap_n_drops_buffer_total{dir="exit",drop="clone_fork"} 0
# HELP falcosecurity_scap_n_drops_buffer_total https://falco.org/docs/metrics/
# TYPE falcosecurity_scap_n_drops_buffer_total counter
falcosecurity_scap_n_drops_buffer_total{dir="enter",drop="execve"} 0
# HELP falcosecurity_scap_n_drops_buffer_total https://falco.org/docs/metrics/
# TYPE falcosecurity_scap_n_drops_buffer_total counter
falcosecurity_scap_n_drops_buffer_total{dir="exit",drop="execve"} 0
# HELP falcosecurity_scap_n_drops_buffer_total https://falco.org/docs/metrics/
# TYPE falcosecurity_scap_n_drops_buffer_total counter
falcosecurity_scap_n_drops_buffer_total{dir="enter",drop="connect"} 0
# HELP falcosecurity_scap_n_drops_buffer_total https://falco.org/docs/metrics/
# TYPE falcosecurity_scap_n_drops_buffer_total counter
falcosecurity_scap_n_drops_buffer_total{dir="exit",drop="connect"} 0
# HELP falcosecurity_scap_n_drops_buffer_total https://falco.org/docs/metrics/
# TYPE falcosecurity_scap_n_drops_buffer_total counter
falcosecurity_scap_n_drops_buffer_total{dir="enter",drop="open"} 0
# HELP falcosecurity_scap_n_drops_buffer_total https://falco.org/docs/metrics/
# TYPE falcosecurity_scap_n_drops_buffer_total counter
falcosecurity_scap_n_drops_buffer_total{dir="exit",drop="open"} 0
# HELP falcosecurity_scap_n_drops_buffer_total https://falco.org/docs/metrics/
# TYPE falcosecurity_scap_n_drops_buffer_total counter
falcosecurity_scap_n_drops_buffer_total{dir="enter",drop="dir_file"} 0
# HELP falcosecurity_scap_n_drops_buffer_total https://falco.org/docs/metrics/
# TYPE falcosecurity_scap_n_drops_buffer_total counter
falcosecurity_scap_n_drops_buffer_total{dir="exit",drop="dir_file"} 0
# HELP falcosecurity_scap_n_drops_buffer_total https://falco.org/docs/metrics/
# TYPE falcosecurity_scap_n_drops_buffer_total counter
falcosecurity_scap_n_drops_buffer_total{dir="enter",drop="other_interest"} 0
# HELP falcosecurity_scap_n_drops_buffer_total https://falco.org/docs/metrics/
# TYPE falcosecurity_scap_n_drops_buffer_total counter
falcosecurity_scap_n_drops_buffer_total{dir="exit",drop="other_interest"} 0
# HELP falcosecurity_scap_n_drops_buffer_total https://falco.org/docs/metrics/
# TYPE falcosecurity_scap_n_drops_buffer_total counter
falcosecurity_scap_n_drops_buffer_total{dir="exit",drop="close"} 0
# HELP falcosecurity_scap_n_drops_buffer_total https://falco.org/docs/metrics/
# TYPE falcosecurity_scap_n_drops_buffer_total counter
falcosecurity_scap_n_drops_buffer_total{dir="exit",drop="proc"} 0
# HELP falcosecurity_scap_n_drops_scratch_map_total https://falco.org/docs/metrics/
# TYPE falcosecurity_scap_n_drops_scratch_map_total counter
falcosecurity_scap_n_drops_scratch_map_total 0
# HELP falcosecurity_scap_n_drops_page_faults_total https://falco.org/docs/metrics/
# TYPE falcosecurity_scap_n_drops_page_faults_total counter
falcosecurity_scap_n_drops_page_faults_total 0
# HELP falcosecurity_scap_n_drops_bug_total https://falco.org/docs/metrics/
# TYPE falcosecurity_scap_n_drops_bug_total counter
falcosecurity_scap_n_drops_bug_total 16
# HELP falcosecurity_scap_n_drops_total https://falco.org/docs/metrics/
# TYPE falcosecurity_scap_n_drops_total counter
falcosecurity_scap_n_drops_total 16
# HELP falcosecurity_falco_cpu_usage_ratio https://falco.org/docs/metrics/
# TYPE falcosecurity_falco_cpu_usage_ratio gauge
falcosecurity_falco_cpu_usage_ratio 0.100000
# HELP falcosecurity_falco_memory_rss_bytes https://falco.org/docs/metrics/
# TYPE falcosecurity_falco_memory_rss_bytes gauge
falcosecurity_falco_memory_rss_bytes 115421184.000000
# HELP falcosecurity_falco_memory_vsz_bytes https://falco.org/docs/metrics/
# TYPE falcosecurity_falco_memory_vsz_bytes gauge
falcosecurity_falco_memory_vsz_bytes 3381981184.000000
# HELP falcosecurity_falco_memory_pss_bytes https://falco.org/docs/metrics/
# TYPE falcosecurity_falco_memory_pss_bytes gauge
falcosecurity_falco_memory_pss_bytes 106614784.000000
# HELP falcosecurity_falco_container_memory_used_bytes https://falco.org/docs/metrics/
# TYPE falcosecurity_falco_container_memory_used_bytes gauge
falcosecurity_falco_container_memory_used_bytes 0.000000
# HELP falcosecurity_falco_host_cpu_usage_ratio https://falco.org/docs/metrics/
# TYPE falcosecurity_falco_host_cpu_usage_ratio gauge
falcosecurity_falco_host_cpu_usage_ratio 0.084000
# HELP falcosecurity_falco_host_memory_used_bytes https://falco.org/docs/metrics/
# TYPE falcosecurity_falco_host_memory_used_bytes gauge
falcosecurity_falco_host_memory_used_bytes 13185093632.000000
# HELP falcosecurity_falco_host_procs_running_total https://falco.org/docs/metrics/
# TYPE falcosecurity_falco_host_procs_running_total gauge
falcosecurity_falco_host_procs_running_total 2
# HELP falcosecurity_falco_host_open_fds_total https://falco.org/docs/metrics/
# TYPE falcosecurity_falco_host_open_fds_total gauge
falcosecurity_falco_host_open_fds_total 21032
# HELP falcosecurity_scap_n_threads_total https://falco.org/docs/metrics/
# TYPE falcosecurity_scap_n_threads_total gauge
falcosecurity_scap_n_threads_total 2407
# HELP falcosecurity_scap_n_fds_total https://falco.org/docs/metrics/
# TYPE falcosecurity_scap_n_fds_total gauge
falcosecurity_scap_n_fds_total 218233
# HELP falcosecurity_scap_n_noncached_fd_lookups_total https://falco.org/docs/metrics/
# TYPE falcosecurity_scap_n_noncached_fd_lookups_total counter
falcosecurity_scap_n_noncached_fd_lookups_total 122148
# HELP falcosecurity_scap_n_cached_fd_lookups_total https://falco.org/docs/metrics/
# TYPE falcosecurity_scap_n_cached_fd_lookups_total counter
falcosecurity_scap_n_cached_fd_lookups_total 214552
# HELP falcosecurity_scap_n_failed_fd_lookups_total https://falco.org/docs/metrics/
# TYPE falcosecurity_scap_n_failed_fd_lookups_total counter
falcosecurity_scap_n_failed_fd_lookups_total 3480
# HELP falcosecurity_scap_n_added_fds_total https://falco.org/docs/metrics/
# TYPE falcosecurity_scap_n_added_fds_total counter
falcosecurity_scap_n_added_fds_total 49706
# HELP falcosecurity_scap_n_removed_fds_total https://falco.org/docs/metrics/
# TYPE falcosecurity_scap_n_removed_fds_total counter
falcosecurity_scap_n_removed_fds_total 41193
# HELP falcosecurity_scap_n_stored_evts_total https://falco.org/docs/metrics/
# TYPE falcosecurity_scap_n_stored_evts_total counter
falcosecurity_scap_n_stored_evts_total 46790
# HELP falcosecurity_scap_n_store_evts_drops_total https://falco.org/docs/metrics/
# TYPE falcosecurity_scap_n_store_evts_drops_total counter
falcosecurity_scap_n_store_evts_drops_total 0
# HELP falcosecurity_scap_n_retrieved_evts_total https://falco.org/docs/metrics/
# TYPE falcosecurity_scap_n_retrieved_evts_total counter
falcosecurity_scap_n_retrieved_evts_total 45570
# HELP falcosecurity_scap_n_retrieve_evts_drops_total https://falco.org/docs/metrics/
# TYPE falcosecurity_scap_n_retrieve_evts_drops_total counter
falcosecurity_scap_n_retrieve_evts_drops_total 39321
# HELP falcosecurity_scap_n_noncached_thread_lookups_total https://falco.org/docs/metrics/
# TYPE falcosecurity_scap_n_noncached_thread_lookups_total counter
falcosecurity_scap_n_noncached_thread_lookups_total 89048
# HELP falcosecurity_scap_n_cached_thread_lookups_total https://falco.org/docs/metrics/
# TYPE falcosecurity_scap_n_cached_thread_lookups_total counter
falcosecurity_scap_n_cached_thread_lookups_total 283785
# HELP falcosecurity_scap_n_failed_thread_lookups_total https://falco.org/docs/metrics/
# TYPE falcosecurity_scap_n_failed_thread_lookups_total counter
falcosecurity_scap_n_failed_thread_lookups_total 397
# HELP falcosecurity_scap_n_added_threads_total https://falco.org/docs/metrics/
# TYPE falcosecurity_scap_n_added_threads_total counter
falcosecurity_scap_n_added_threads_total 2625
# HELP falcosecurity_scap_n_removed_threads_total https://falco.org/docs/metrics/
# TYPE falcosecurity_scap_n_removed_threads_total counter
falcosecurity_scap_n_removed_threads_total 218
# HELP falcosecurity_scap_n_drops_full_threadtable_total https://falco.org/docs/metrics/
# TYPE falcosecurity_scap_n_drops_full_threadtable_total counter
falcosecurity_scap_n_drops_full_threadtable_total 0
# HELP falcosecurity_plugins_container_n_containers_total https://falco.org/docs/metrics/
# TYPE falcosecurity_plugins_container_n_containers_total counter
falcosecurity_plugins_container_n_containers_total 5
# HELP falcosecurity_plugins_container_n_missing_container_images_total https://falco.org/docs/metrics/
# TYPE falcosecurity_plugins_container_n_missing_container_images_total counter
falcosecurity_plugins_container_n_missing_container_images_total 0
# HELP falcosecurity_falco_kernel_release_info https://falco.org/docs/metrics/
# TYPE falcosecurity_falco_kernel_release_info gauge
falcosecurity_falco_kernel_release_info{kernel_release="6.8.0-60-generic"} 1
# HELP falcosecurity_evt_hostname_info https://falco.org/docs/metrics/
# TYPE falcosecurity_evt_hostname_info gauge
falcosecurity_evt_hostname_info{hostname="federico.dipierro"} 1
# HELP falcosecurity_falco_outputs_queue_num_drops_total https://falco.org/docs/metrics/
# TYPE falcosecurity_falco_outputs_queue_num_drops_total counter
falcosecurity_falco_outputs_queue_num_drops_total 0
# HELP falcosecurity_falco_reload_timestamp_nanoseconds https://falco.org/docs/metrics/
# TYPE falcosecurity_falco_reload_timestamp_nanoseconds gauge
falcosecurity_falco_reload_timestamp_nanoseconds 1748339331904725422
# HELP falcosecurity_falco_start_timestamp_nanoseconds https://falco.org/docs/metrics/
# TYPE falcosecurity_falco_start_timestamp_nanoseconds gauge
falcosecurity_falco_start_timestamp_nanoseconds 1748339331903745044
# HELP falcosecurity_falco_duration_seconds_total https://falco.org/docs/metrics/
# TYPE falcosecurity_falco_duration_seconds_total counter
falcosecurity_falco_duration_seconds_total 13
# HELP falcosecurity_falco_host_boot_timestamp_nanoseconds https://falco.org/docs/metrics/
# TYPE falcosecurity_falco_host_boot_timestamp_nanoseconds gauge
falcosecurity_falco_host_boot_timestamp_nanoseconds 1748328212000000000
# HELP falcosecurity_falco_host_num_cpus_total https://falco.org/docs/metrics/
# TYPE falcosecurity_falco_host_num_cpus_total gauge
falcosecurity_falco_host_num_cpus_total 20
```

There are no duplicated (tested with bash with a Falco run with `bpf` syscall source + `dummy` source enabled, + container plugin loaded with the syscall source)

FedeDP commented:
Attached a metrics from 0.40.0 from same configuration (without the container plugin obviously because 0.40.0 did not support it): 
[metrics_040.txt](https://github.com/user-attachments/files/20455403/metrics_040.txt)
We can see that there are multiple duplicated entries, as stated by the #3539 issue.

poiana commented:
LGTM label has been added.  <details>Git tree hash: 3bd15a0d5982d834506b4079176ab6c09f4a162b</details>

FedeDP commented:
@sgaist great catch! I will fix them tomorrow :) thanks for the review!
/hold

poiana commented:
New changes are detected. LGTM label has been removed.

FedeDP commented on userspace/falco/falco_metrics.h:
```diff
@@ -26,6 +26,16 @@ struct state;
-	static const std::string content_type;
-	static std::string to_text(const falco::app::state& state);
+	static const std::string content_type_prometheus;
+	static std::string to_text_prometheus(const falco::app::state& state);
```
I think `to_text_prometheus` is easy to get if we'll ever add new `to_text_foo`, that's why i chose it :)

FedeDP commented:
/unhold
